### PR TITLE
Microsoft.CodeTransparency 2024-01-11-preview Dataplane APIs

### PR DESCRIPTION
**Pull request by vimauro**
# Data Plane API - Pull Request

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*API Spec engagement record*](https://aka.ms/azsdk/onboarding/restapischedule).

* Link to API Spec engagement record issue:

Is this review for (select one):

- [x] a private preview
- [ ] a public preview
- [ ] GA release

### Change Scope

<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous API Spec document (if applicable), and the root paths that have been updated. </sup>
* Design Document:
* Previous API Spec Doc:
* Updated paths:

### Viewing API changes

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

### Suppressing failures

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[Swagger-Suppression-Process](https://aka.ms/pr-suppressions) 
to get approval.

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azsdk/onboarding/restapischedule) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azsdk/onboarding/restapischedule).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling

 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://github.com/Azure/azure-api-style-guide/blob/main/README.md)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications

 * [Azure REST API Guidelines](https://aka.ms/azapi/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/azapi/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links

 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

<details>
  <summary>Checks stuck in `queued` state?</summary>
If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.  
</details>

### CODE REVIEW

**Comment by weidongxu-microsoft** [comment]
Please see if you need to schedule a review
https://aka.ms/azsdk/onboarding/restapischedule

**Comment by ivarprudnikov** [comment]
We have done the review for this already in the private repo. But it needs to change a bit due to SDK feedback https://apiview.dev/Assemblies/Review/02c4a28fc696407e8b3242a7210e74df

**Comment by weidongxu-microsoft** [comment]
Could you give us a link to the review/approved private repo PR? If possible also since which commit is your new changes.

This way, it would be much easier for us the just review the changed part, instead of all the files (that appears as new in this PR).

**Comment by ivarprudnikov** [comment]
The PR in the private repo that was already discussed in the call https://github.com/Azure/azure-rest-api-specs-pr/pull/16579 

There were more changes after the fact to make sure generator changes the code as required. This was not merged because the PR is against the `main` which is not possible in the private repo.

**Comment by weidongxu-microsoft** [comment]
> The PR in the private repo that was already discussed in the call [Azure/azure-rest-api-specs-pr#16579](https://github.com/Azure/azure-rest-api-specs-pr/pull/16579)
> 
> There were more changes after the fact to make sure generator changes the code as required. This was not merged because the PR is against the `main` which is not possible in the private repo.
> 
> Had to do some tweaks (flatten the api instead of using interfaces) in the past day or two because of the comments in the API review https://apiview.dev/Assemblies/Review/02c4a28fc696407e8b3242a7210e74df

Got it. Note that the PR on private repo is not approved. So we probably just carry the discussion here https://github.com/Azure/azure-rest-api-specs-pr/issues/16580
(may sure you either address them, or comment about the reason why you cannot follow the advice)

The review can go async without another meeting, as we already had a review meeting and issues listed. Currently this is a private preview.

**Comment by weidongxu-microsoft** [comment]
On the CI, make sure you always had at least "TypeSpec Validation" pass. (there is other required CIs, but since you write tsp, this one is the most important)

**Comment by ivarprudnikov** [comment]
> On the CI, make sure you always had at least "TypeSpec Validation" pass. (there is other required CIs, but since you write tsp, this one is the most important)

It was done already but I think the head changed and a rebase is necessary to be able to reproduce the issue

> The review can go async without another meeting, as we already had a review meeting and issues listed. Currently this is a private preview.

Could you please explain what you need from us now for this PR to be approved? Where do we do this async review?

**Comment by ivarprudnikov** [comment]
Just to make sure I understand, this is a private preview we are talking about and we agreed this is good as it is and we'll resolve any outstanding issues in the next api version. Still though I'll try to "respond" to those comments in this PR.

## The ask is your response from last meeting.

Comments to the summary in https://github.com/Azure/azure-rest-api-specs-pr/issues/16580#issuecomment-1919601365 added below:

> This doesn't use many Azure.Core types e.g., there's no actual "entries" resources.
- Correct, an entry is a binary that we store, it is part of the RFC
> Naming convention of operations isn't standard, partly because of the point above.
- Correct, part of the RFC
>However, it's critical to note they are implementing this RFC: https://datatracker.ietf.org/doc/draft-ietf-scitt-architecture/. This is to be compatible with other implementations in the industry.
- We are not on the latest draft yet, but endpoints can be seen in v4 https://datatracker.ietf.org/doc/draft-ietf-scitt-architecture/04/ in section "8 Transparency Service API"
>They do add some endpoints atop that spec, and there's where we should follow Azure API guidelines as much as possible.
- see below
>They do inherit some endpoints and backend implementation of Confidential Ledger - they build on the same framework.
- Correction, we inherit from CCF (Confidential consortium Framework) that went GA recently and it is almost identical to the ones in Confidential Ledger as they inherited the same endpoints.
>They also have some control plane-like endpoints they don't document but are not currently going through ARM.
- We rely on MCCF control plane at the moment

## API review comments

> E.g. I didn't see response to comments  https://apiview.dev/Assemblies/Conversation/dc124dc803df45ac8d694409b75b352f

Answered the comments but nobody really asked for them to be answered and we agreed that the current version is good as it is.

## API surface

> Particularly, let us know which API is from https://datatracker.ietf.org/doc/draft-ietf-scitt-architecture/, and which is added atop of that RFC.

Read through [Draft 4](https://datatracker.ietf.org/doc/draft-ietf-scitt-architecture/04/) section "8.  Transparency Service API", the endpoints are:

- `POST <Base URL>/entries`
- `GET <Base URL>/operations/<Operation ID>`
- `GET <Base URL>/entries/<Entry ID>`
- `GET <Base URL>/entries/<Entry ID>/receipt`

Additional endpoints (not the inherited CCF ones):
- `GET /.well-known/did.json` based on spec https://w3c-ccg.github.io/did-method-web/
- `GET /parameters` gets a json blob representing the service state (id, public key, supported algorithms)
- `GET /configuration` gets a json blob that represents current service configuration
- `GET /version` gets some json that shows the version of "scitt" the service is running

Inherited endpoints from CCF, the same can be found in Confidential Ledger and in other applications that launch in MCCF in Azure:
- https://microsoft.github.io/CCF/main/use_apps/rpc_api.html
- https://microsoft.github.io/CCF/main/operations/operator_rpc_api.html
- https://microsoft.github.io/CCF/main/governance/member_rpc_api.html

**Comment by weidongxu-microsoft** [comment]
Thanks for the summary.

It appears most discussion is about the inconsistency of property naming within the service. E.g. it is camelCase in `ParametersResult`, but snake_case in `CodeTransparencyConfiguration`. Both of the API is additional endpoint.

The RFC API "Transparency Service API" and `DidDocument` is also camelCase.

I guess the question is whether we allow this inconsistency into private preview, and ask for fix in public preview.

I've added Heath to the reviewers.

**Comment by ivarprudnikov** [comment]
@weidongxu-microsoft @heaths Is there anything you are waiting for? Can we approve this PR and move forward and allow other teams to test the Private Preview version?

**Review by heaths — APPROVED** [review]
All the changes I requested seem to have been addressed. Thank you!

**Comment by heaths** [comment]
@weidongxu-microsoft if you've reviewed this, can you sign off and add the appropriate label (since we need 2)?

**Review by weidongxu-microsoft — APPROVED** [review]

File: specification/confidentialledger/Microsoft.CodeTransparency/models.tsp
Comment (weidongxu-microsoft) [inline specification/confidentialledger/Microsoft.CodeTransparency/models.tsp:275]: Please make sure you would fix the casing in next api-version (or provide a convincing reason that it had to stay snake_case).

File: specification/confidentialledger/Microsoft.CodeTransparency/entries.tsp
Comment (weidongxu-microsoft) [inline specification/confidentialledger/Microsoft.CodeTransparency/entries.tsp:65]: I assume the lint error of `DefaultResponse` refer to here (you suppressed it in readme).

If you check the Swagger, the API would only document 200 and 503. I assume the API could have more than these status code when it fails.
It is not a blocker on private preview, but could you see if you can use e.g. `ServiceUnavailableError | Azure.Core.Foundations.ErrorResponse`

File: specification/confidentialledger/Microsoft.CodeTransparency/models.tsp
Comment (weidongxu-microsoft) [inline specification/confidentialledger/Microsoft.CodeTransparency/models.tsp:147]: Please use `union` for extensible enum.

Context https://github.com/Azure/azure-rest-api-specs/pull/27687

File: specification/confidentialledger/Microsoft.CodeTransparency/models.tsp
Comment (ivarprudnikov) [reply specification/confidentialledger/Microsoft.CodeTransparency/models.tsp:272]: I've raised it as a separate issue and we'll track and fix it in a later version https://github.com/microsoft/scitt-ccf-ledger/issues/187

File: specification/confidentialledger/Microsoft.CodeTransparency/entries.tsp
Comment (ivarprudnikov) [reply specification/confidentialledger/Microsoft.CodeTransparency/entries.tsp]: Added default response, thanks for the suggestion on how to achieve it

**Pull request merged**
